crates/ruff_linter/src/rules/flake8_bugbear/rules/loop_iterator_mutated.rs

```diff
@@ -0,0 +1,185 @@
+                    return "".into();
+                }
+                _ => {
+                    inner.push_str(".");
+                    inner.push_str(attr);
+                    return inner;
+                }
+            }
+        }
+        Expr::Call(ExprCall { range: _, func, .. }) => {
+            return _to_name_str(func);
+        }
+        _ => {
+            return "".into();
+        }
+    }
+}
+// B038
+pub(crate) fn loop_iterator_mutated(checker: &mut Checker, stmt_for: &StmtFor) {
+    let StmtFor {
+        target: _,
+        iter,
+        body,
+        orelse: _,
+        is_async: _,
+        range: _,
+    } = stmt_for;
+    let name;
+
+    match iter.as_ref() {
+        Expr::Name(ExprName { .. }) => {
+            name = _to_name_str(iter.as_ref());
+        }
+        Expr::Attribute(ExprAttribute { .. }) => {
+            name = _to_name_str(iter.as_ref());
+        }
+        _ => {
+            println!("Shouldn't happen");
+            return;
+        }
+    }
+    let mut visitor = LoopMutationsVisitor {
+        name: &name,
+        mutations: Vec::new(),
+    };
+    visitor.visit_body(body);
+    for mutation in visitor.mutations {
+        checker
+            .diagnostics
+            .push(Diagnostic::new(LoopIteratorMutated {}, mutation.range()));
+    }
+}
+struct LoopMutationsVisitor<'a> {
+    name: &'a String,
+    mutations: Vec<Box<dyn Ranged + 'a>>,
+}
+
+/// `Visitor` to collect all used identifiers in a statement.
+impl<'a> Visitor<'a> for LoopMutationsVisitor<'a> {
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        match stmt {
+            Stmt::Delete(StmtDelete { range, targets }) => {
+                for target in targets {
+                    let name;
+                    match target {
+                        Expr::Subscript(ExprSubscript {
+                            range: _,
+                            value,
+                            slice: _,
+                            ctx: _,
+                        }) => {
+                            name = _to_name_str(value);
+                        }
+
+                        Expr::Attribute(_) | Expr::Name(_) => {
+                            name = _to_name_str(target);
+                        }
+                        _ => {
+                            name = String::new();
+                            visitor::walk_expr(self, target);
+                        }
+                    }
+                    if self.name.eq(&name) {
+                        self.mutations.push(Box::new(range));
+                    }
+                }
+            }
+            _ => {
+                visitor::walk_stmt(self, stmt);
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match expr {
+            Expr::Name(ExprName { range: _, id, ctx }) => {
+                if self.name.eq(id) {
+                    match ctx {
+                        ExprContext::Del => {
```

Comment:
Do we need this, if we're already look at `Stmt::Delete` above?

Separately, what about `ExprContext::Store`?

Reply:
We will need `ExprContext::Store`, and quite some more additions (see https://github.com/PyCQA/flake8-bugbear/issues/454), but I don't want to pollute the current review with this for your ease. 

I'd add all these improvements later on, once this PR would be ready.